plugins/inputs/cloud_pubsub/pubsub.go

```diff
@@ -170,30 +177,39 @@ func (ps *PubSub) startReceiver(parentCtx context.Context) error {
 
 // onMessage handles parsing and adding a received message to the accumulator.
 func (ps *PubSub) onMessage(ctx context.Context, msg message) error {
+	defer msg.Ack()
 	if ps.MaxMessageLen > 0 && len(msg.Data()) > ps.MaxMessageLen {
-		msg.Ack()
 		return fmt.Errorf("message longer than max_message_len (%d > %d)", len(msg.Data()), ps.MaxMessageLen)
 	}
 
+	// This function is called concurrently, but the decoder cannot.
+	ps.decoderMutex.Lock()
+	b, err := ps.decoder.Decode(msg.Data())
+	ps.decoderMutex.Unlock()
+	if err != nil {
+		return fmt.Errorf("unable to decode message: %v", err)
+	}
 	var data []byte
+	if ps.ContentEncoding == "gzip" {
+		data = make([]byte, len(b))
+		copy(data, b)
```

Comment:
why does gzip need a copy of this?
seems like this whole conditional could just be `data = b`

Reply:
IMHO the copy is needed because GzipEncoder [Encode ](https://github.com/influxdata/telegraf/blob/master/internal/content_coding.go#L125) is returning [buffer.Bytes() ](https://pkg.go.dev/bytes#Buffer.Bytes)which is a [slice](https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/bytes/buffer.go;l=54) and if we multiple goroutines are modifying the underlying buffer, it starts acting weirdly due to race condition. 
We were initially not doing the copy and started getting bad data in our influxDB and also these errors in telegraf 
```
2021-07-28T10:14:18Z E! [inputs.cloud_pubsub] Error in plugin: unable to add message from subscription metrics-platform: unable to parse decoded message: metric parse error: expected field at 309:327: "luster_id=140046,container_image=us.gcr.io/wp-engine-development/memcached,container_name=memcached,container_version=24,cpu=cpu1,datacenter=gce-us-east1,host=host1,innodb_buffer_reduce=0,offering=shared,platform=classic,region=us-east1,server_version=20.10.7,ws_external_hats=False usage_total=3510062015i 1627467241000000000"
2021-07-28T10:14:18Z E! [inputs.cloud_pubsub] Error in plugin: unable to add message from subscription metrics-platform: unable to parse decoded message: metric parse error: expected tag at 507:177: ""
```
and this resulted in the following
<img width="727" alt="error" src="https://user-images.githubusercontent.com/49992258/127455243-132b9ca1-d91e-4990-a949-97916fd7a177.png">
 BTW I also saw we handle the buffer in similar way in [Influx serialization code](https://github.com/influxdata/telegraf/blob/1ea5c20aadc55863ce1acd49051e6d6db03b4d60/plugins/serializers/influx/influx.go#L105-L106). 
 By using similar code like above solved our problem in production.

Reply:
I think we're moving away from copies in that code as well. If you can find a way to make it goroutine-safe without the copy, it will be significantly faster.

Reply:
I'm concerned there's a more fundamental issue here. I think what I'd prefer to see is the whole oNMessage function with a mutex, then removing the decode mutex and removing the data copy. You should see all concurrency issues go away.